cpu/stm32/periph/timer.c

```diff
@@ -36,6 +36,163 @@ static inline TIM_TypeDef *dev(tim_t tim)
     return timer_config[tim].dev;
 }
 
+#ifdef MODULE_PERIPH_TIMER_PERIODIC
+
+/**
+ * @brief   Helper macro to get channel bit in timer/channel bitmap
+ */
+#define CHAN_BIT(tim, chan) (1 << chan) << (TIMER_CHANNEL_NUMOF * (tim & 1))
+
+/**
+ * @brief   Bitmap for compare channel disable after match
+ */
+static uint8_t _oneshot[(TIMER_NUMOF+1)/2];
+
+/**
+ * @brief   Bitmap to reset counter on channel match
+ */
+static uint8_t _reset_on_match[(TIMER_NUMOF+1)/2];
+
+/**
+ * @brief   Ignore one IRQ of the timer
+ */
+static uint16_t _ignore_irq;
+
+/**
+ * @brief   Clear interrupt enable after the interrupt has fired
+ */
+static inline void set_oneshot(tim_t tim, int chan)
+{
+    _oneshot[tim >> 1] |= CHAN_BIT(tim, chan);
+}
+
+/**
+ * @brief   Enable interrupt with every wrap-around of the timer
+ */
+static inline void clear_oneshot(tim_t tim, int chan)
+{
+    _oneshot[tim >> 1] &= ~CHAN_BIT(tim, chan);
+}
+
+static inline bool is_oneshot(tim_t tim, int chan)
+{
+    return _oneshot[tim >> 1] & CHAN_BIT(tim, chan);
+}
+
+/**
+ * @brief   ignore the first IRQ
+ */
+static inline void set_ignore_irq(tim_t tim)
+{
+    _ignore_irq |= 1 << tim;
+}
+
+/**
+ * @brief   don't ignore IRQ anymore
+ */
+static inline void clear_ignore_irq(tim_t tim)
+{
+    _ignore_irq &= ~(1 << tim);
+}
+
+static inline bool is_ignore_irq(tim_t tim)
+{
+    return _ignore_irq & (1 << tim);
+}
+
+static void _set_arr(tim_t tim);
+
+/**
+ * @brief   Reset the timer on channel match
+ */
+static inline void set_reset_on_match(tim_t tim, int chan)
+{
+    _reset_on_match[tim >> 1] |= CHAN_BIT(tim, chan);
+    _set_arr(tim);
+}
+
+/**
+ * @brief   Keep counting on channel match
+ */
+static inline void clear_reset_on_match(tim_t tim, int chan)
+{
+    _reset_on_match[tim >> 1] &= ~CHAN_BIT(tim, chan);
+    _set_arr(tim);
+}
+
+static inline bool is_reset_on_match(tim_t tim, int chan)
+{
+    return _reset_on_match[tim >> 1] & CHAN_BIT(tim, chan);
+}
+
+/**
+ * @brief   Set timer TOP to the lowest reset on match channel
+ */
+static void _set_arr(tim_t tim)
+{
+    uint32_t dir = dev(tim)->DIER;
+    uint32_t arr = timer_config[tim].max;
+
+    /* find the lowest reset value */
+    for (unsigned int i = 0; i < TIMER_CHANNEL_NUMOF; i++) {
+
+        /* skip disabled channels */
+        if ((dir & (TIM_DIER_CC1IE << i)) == 0) {
+            continue;
+        }
+
+        uint32_t cc = TIM_CHAN(tim, i);
+        if (is_reset_on_match(tim, i) && cc < arr) {
+            arr = cc;
+        }
+    }
+
+    dev(tim)->ARR = arr;
+}
+#else /* !MODULE_PERIPH_TIMER_PERIODIC */
```

Comment:
I am guessing this #else is just so you don't have to #ifdef the module everywhere?

Reply:
Yes, the dummy functions are there to avoid unneeded overhead if the new function is not used.